### ISSUE TITLE
array implementation of into_iter() causes stack overflow

### ISSUE THREAD

**Issue by mbaulch** [rust-lang/rust#111598]
The following code (debug build on Android Aarch64 with Termux) aborts with a "Segmentation fault":

```rust
const NUM_FOOS: usize = 200000;
const FOOS: [u32; NUM_FOOS] = [0; NUM_FOOS];

fn main() {
    for f in FOOS {}
}
```

On Linux x86-64 a stack overflow is reported. Then again, is it reasonable for this code to cause a stack overflow? I'm inclined to think it shouldn't.

If we call FOOS.iter() rather than rely on into_iter() being called implicitly, the issue disappears. Perhaps of note is that unsafe code (which, it seems, makes a number of assumptions) is used in the array implementation of into_iter().

### Meta

`rustc --version --verbose`:
```
rustc 1.69.0 (84c898d65 2023-04-16) (built from a source tarball)
```

<details><summary>Backtrace</summary>
<p>

```
#0  0x000000555574e544 in core::array::iter::{impl#0}::into_iter<u32, 500000> (self=...)
    at /home/builder/.termux-build/rust/src/library/core/src/array/iter.rs:54
#1  0x000000555574eb4c in as_bin::main () at src/main.rs:5
```

</p>
</details>

**Comment by clubby789** [comment]
The issue is that `into_iter` takes an owned value, which means `FOOS` is constructed on the stack - it's much too large and overflows. `.iter()` takes a reference, which means that a large static array is instead compiled into the binary and a reference to this is taken.
If you make `FOOS` a static, it will always be embedded into the binary so you can avoid blowing up the stack.

**Comment by mbaulch** [comment]
Thanks for your explanation @clubby789. That makes perfect sense, though I admit it does feel like a slightly sharp and perhaps unexpected edge to the language or implementation.

I may have erred by oversimplifying the code that caused my original issue. The following variation causes the segfault to be triggered within core::mem::transmute_copy, which, it seems is called by into_iter in this case:

```rust
const NUM_A: usize = 11;

#[derive(Clone, Copy)]
struct Foo {
    a: [&'static str; NUM_A],
}

const NUM_FOO: usize = 1500;

const FOOS: [Foo; NUM_FOO] = [Foo { a: [""; NUM_A] }; NUM_FOO];

#[cfg(test)]
mod tests {
    #[test]
    fn test_foo() {
        for f in super::FOOS {}
    }
}
```
<details><summary>Backtrace</summary>
<p>

```
#0  0x00000055555e6914 in core::mem::transmute_copy<[lang::Foo; 1500], [core::mem::maybe_uninit::MaybeUninit<lang::Foo>; 1500]> () at /home/builder/.termux-build/rust/src/library/core/src/mem/mod.rs:1047
#1  0x00000055555e657c in core::array::iter::{impl#0}::into_iter<lang::Foo, 1500> ()
    at /home/builder/.termux-build/rust/src/library/core/src/array/iter.rs:73
#2  0x00000055555e6b90 in as_bin::tests::test_foo () at src/lib.rs:16
#3  0x00000055555e77d4 in as_bin::tests::test_foo::{closure#0} () at src/lib.rs:15
#4  0x00000055555e6d00 in core::ops::function::FnOnce::call_once<lang::tests::test_foo::{closure_env#0}, ()> ()
    at /home/builder/.termux-build/rust/src/library/core/src/ops/function.rs:250
#5  0x00000055556062b8 in test::__rust_begin_short_backtrace::h5295331c10838615 ()
#6  0x00000055556054b0 in test::run_test::run_test_inner::{{closure}}::hdb6e5fbd252316c8 ()
#7  0x00000055555e8028 in std::sys_common::backtrace::__rust_begin_short_backtrace::h62188ba109d57d3c ()
#8  0x0000005555609b24 in core::ops::function::FnOnce::call_once{{vtable.shim}}::hcb784459aeff44a4 ()
#9  0x00000055556282d4 in std::sys::unix::thread::Thread::new::thread_start::h573c53a4ca02e775 ()
#10 0x0000007ff2b4b72c in __pthread_start(void*) () from /apex/com.android.runtime/lib64/bionic/libc.so
#11 0x0000007ff2ae5020 in __start_thread () from /apex/com.android.runtime/lib64/bionic/libc.so
```

</p>
</details>

It seems that the stack size is reduced for code run within the test harness. These numbers result in different behaviour when run in a standard binary with a main() function, though my guess is it'd be possible to replicate the same behaviour by adjusting them appropriately. If NUM_FOO is set too large, the crash occurs in into_iter as in the original code above.

Regarding my use case, I was under the mistaken impression that const functions can only be used to initialise const variables, but it seems that they can be used to initialise static variables too. Programming with const functions is a rapidly developing feature, so perhaps it's understandable that the documentation around this isn't as carefully curated.

My feeling is that it would be good if the compiler issued a warning if one tries to assign a large struct (or array) to a const variable and recommends it be changed to static.  I'm considering whether to close this issue and reopen with a feature request instead, though if anyone has any further comments, I'd very much welcome them.

**Comment by mbaulch** [comment]
I've just tested @clubby789's suggestion to replace "const FOOS" with "static FOOS" and it seems to not have any effect on the issue. Presumably the static FOOS ends up being copied onto the stack as part of the function call either way. Sorry, I should have thought to check this originally.

**Comment by mbaulch** [comment]
I'll close this issue, and perhaps just conclude that in general passing an array by valueーor indeed another data structure (such as a struct or enum) containing itーcarries significant risks in terms of exhausting stack memory. This is particularly exacerbated in the case that the array's size is not fixed: due to const generics, for example.

It seems that compiler or language features are required to provide proper guardrails for these issues. Hopefully one day soon! This feels like one of the exceedingly rare areas where Rust fails to live up to its promises.
